src/System.Management.Automation/namespaces/FileSystemProvider.cs

```diff
@@ -8216,28 +8230,55 @@ internal static bool IsReparsePoint(FileSystemInfo fileInfo)
             return fileInfo.Attributes.HasFlag(System.IO.FileAttributes.ReparsePoint);
         }
 
-        internal static bool IsReparsePointWithTarget(FileSystemInfo fileInfo)
+        internal static bool IsReparsePointLikeSymlink(FileSystemInfo fileInfo)
         {
-            if (!IsReparsePoint(fileInfo))
+#if UNIX
+            // Reparse point on Unix is a symlink.
+            return IsReparsePoint(fileInfo);
+#else
+            if (InternalTestHooks.OneDriveTestOn && fileInfo.Name == InternalTestHooks.OneDriveTestSymlinkName)
             {
-                return false;
+                return !InternalTestHooks.OneDriveTestRecurseOn;
             }
-#if !UNIX
-            // It is a reparse point and we should check some reparse point tags.
-            var data = new WIN32_FIND_DATA();
-            using (var handle = FindFirstFileEx(fileInfo.FullName, FINDEX_INFO_LEVELS.FindExInfoBasic, ref data, FINDEX_SEARCH_OPS.FindExSearchNameMatch, IntPtr.Zero, 0))
+
+            WIN32_FIND_DATA data = default;
+            string fullPath = Path.TrimEndingDirectorySeparator(fileInfo.FullName);
+            if (fullPath.Length > MAX_PATH)
+            {
+                fullPath = PathUtils.EnsureExtendedPrefix(fullPath);
+            }
+
+            using (var handle = FindFirstFileEx(fullPath, FINDEX_INFO_LEVELS.FindExInfoBasic, ref data, FINDEX_SEARCH_OPS.FindExSearchNameMatch, IntPtr.Zero, 0))
             {
+                if (handle.IsInvalid)
+                {
+                    // Our handle could be invalidated by something else touching the filesystem,
+                    // so ensure we deal with that possibility here
+                    int lastError = Marshal.GetLastWin32Error();
+                    throw new Win32Exception(lastError);
```

Comment:
IIRC by omiting all arguments in the constructor `Win32Exception` does the `Marshal.GetLastWin32Error()` call. The current behaviour here ignores an error as well, this might introduce a failure that is currently ignored for this particular check.

Reply:
Sorry I don't understand your comment. Please reword.

Reply:
Doing `throw new Win32Exception(Marshal.GetLastWin32Error());` is the same as just `throw new WIn32Exception();`. The empty constructor gets the last win32 error itself if you don't supply it.

Reply:
Thanks for clarify! The pattern comes from history - 40 times in our code. If we want to change this we should fix all in follow PR.

Reply:
@rjmholt If you agree to review :-) I will fix all these patterns in follow PR.